replica/table.cc

```diff
@@ -334,13 +334,27 @@ mutation_reader table::make_streaming_reader(schema_ptr schema, reader_permit pe
 
 mutation_reader table::make_streaming_reader(schema_ptr schema, reader_permit permit, const dht::partition_range& range,
         lw_shared_ptr<sstables::sstable_set> sstables, gc_clock::time_point compaction_time) const {
-    auto& slice = schema->full_slice();
+    auto reader = mutation_reader{nullptr};
     auto trace_state = tracing::trace_state_ptr();
-    const auto fwd = streamed_mutation::forwarding::no;
-    const auto fwd_mr = mutation_reader::forwarding::no;
+    if (range.is_full()) {
+        reader = sstables->make_crawling_reader(
+            std::move(schema),
+            std::move(permit),
+            std::move(trace_state),
+            sstables::default_read_monitor_generator());
+    } else {
+        auto& slice = schema->full_slice();
+        const auto fwd = streamed_mutation::forwarding::no;
+        const auto fwd_mr = mutation_reader::forwarding::no;
+        reader = sstables->make_range_sstable_reader(
+            std::move(schema),
+            std::move(permit),
+            range,
+            slice,
+            std::move(trace_state), fwd, fwd_mr);
```

Comment:
Maybe we should move this to sstable_set. Why should the caller know about these optimizations?

How does this interact with evicting readers due to memory constraints? If we have to re-materialize a crawling reader, how will that work without the index?

How does it work for non-equal-shard source and destination?

@denesb please rename/document make_crawling_reader, the name is too obscure.

Reply:
We can do a simple rename: crawling_reader -> full_scan_reader.

Indeed we can hide this logic and simply instantiate the `full_scan_reader`, every time the caller passes the full range to `make_reader()`. Eviction will just work: if the reader is evicted, a non-full range will be passed when it is re-created, and then we simply create the regular reader.

Reply:
> Maybe we should move this to sstable_set. Why should the caller know about these optimizations?

added a new member function to `sstable_set` instead of dispatching based on range in the `table` class.

> 
> How does this interact with evicting readers due to memory constraints? If we have to re-materialize a crawling reader, how will that work without the index?

probably i'm be missing something, but is this in the scope of this change? @denesb could you shed some lights  on this?

> 
> How does it work for non-equal-shard source and destination?

load-n-stream reads the entire content from sstables, so i don't think the question applies here.  in this change, we only change the behavior when loading the full range for vnodes.

> 
> @denesb please rename/document make_crawling_reader, the name is too obscure.

#20185 should help as well.